### PR TITLE
Add ContainersReady condition into Pod Status

### PR DESCRIPTION
**Pull request by freehan**
**Last 3 commits are new**

Follow up PR of: https://github.com/kubernetes/kubernetes/pull/64057 and https://github.com/kubernetes/kubernetes/pull/64344

Have a single PR for adding ContainersReady per https://github.com/kubernetes/kubernetes/pull/64344#issuecomment-394038384

```release-note
Introduce ContainersReady condition in Pod Status
```

/assign yujuhong for review
/assign thockin for the tiny API change

### CODE REVIEW

File: staging/src/k8s.io/api/core/v1/generated.proto [hunk 1]
```diff
@@ -3052,6 +3057,13 @@ message PodSpec {
   // configuration based on DNSPolicy.
   // +optional
   optional PodDNSConfig dnsConfig = 26;
+
+  // If specified, all readiness gates will be evaluated for pod readiness.
+  // A pod is ready when all its containers are ready AND
+  // all conditions specified in the readiness gates have status equal to "True"
+  // More info: https://github.com/kubernetes/community/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md
+  // +optional
+  repeated PodReadinessGate readinessGates = 28;
```
Comment (dixudx) [inline staging/src/k8s.io/api/core/v1/generated.proto]: change to 27?

File: staging/src/k8s.io/api/core/v1/types.go [hunk 2]
```diff
@@ -2796,6 +2803,13 @@ type PodSpec struct {
 	// configuration based on DNSPolicy.
 	// +optional
 	DNSConfig *PodDNSConfig `json:"dnsConfig,omitempty" protobuf:"bytes,26,opt,name=dnsConfig"`
+
+	// If specified, all readiness gates will be evaluated for pod readiness.
+	// A pod is ready when all its containers are ready AND
+	// all conditions specified in the readiness gates have status equal to "True"
+	// More info: https://github.com/kubernetes/community/blob/master/keps/sig-network/0007-pod-ready%2B%2B.md
+	// +optional
+	ReadinessGates []PodReadinessGate `json:"readinessGates,omitempty" protobuf:"bytes,28,opt,name=readinessGates"`
```
Comment (dixudx) [inline staging/src/k8s.io/api/core/v1/types.go]: change 28 to 27?

File: pkg/kubelet/kubelet_pods.go [hunk 3]
```diff
@@ -1407,6 +1408,11 @@ func (kl *Kubelet) convertStatusToAPIStatus(pod *v1.Pod, podStatus *kubecontaine
 		true,
 	)
 
+	for _, c := range pod.Status.Conditions {
```
Comment (dixudx) [inline pkg/kubelet/kubelet_pods.go]: Better add a comment for this.

Like, "preserve non-system condition types". WDYT?

**Review by dixudx — COMMENTED** [review]
Mostly lgtm. Some nits.

Moreover, `lastTransitionTime` is missing for new PodConditionType `ContainersReady`. You have to
add `updateLastTransitionTime(&status, &oldStatus, v1.ContainersReady)` in method `updateStatusInternal` (pkg/kubelet/status/status_manager.go).

File: pkg/api/v1/pod/util.go [hunk 4]
```diff
@@ -257,9 +257,18 @@ func GetPodCondition(status *v1.PodStatus, conditionType v1.PodConditionType) (i
 	if status == nil {
 		return -1, nil
 	}
-	for i := range status.Conditions {
-		if status.Conditions[i].Type == conditionType {
-			return i, &status.Conditions[i]
+	return GetPodConditionFromList(status.Conditions, conditionType)
+}
+
+// GetPodConditionFromList extracts the provided condition from the given list of condition and returns that.
```
Comment (yujuhong) [inline pkg/api/v1/pod/util.go]: s/and returns that./and returns the index of the condition and the condition.

File: pkg/api/v1/pod/util.go [hunk 5]
```diff
@@ -257,9 +257,18 @@ func GetPodCondition(status *v1.PodStatus, conditionType v1.PodConditionType) (i
...
+// Returns nil and -1 if the condition is not present, and the index of the located condition.
```
Comment (yujuhong) [inline pkg/api/v1/pod/util.go]: s/Returns nil and -1/Returns -1 and nil
Remove `, and the index of the located condition.`

File: pkg/apis/core/types.go [hunk 6]
```diff
@@ -2068,6 +2068,8 @@ const (
 	// PodReasonUnschedulable reason in PodScheduled PodCondition means that the scheduler
 	// can't schedule the pod right now, for example due to insufficient resources in the cluster.
 	PodReasonUnschedulable = "Unschedulable"
+	// ContainersReady indicates whether all containers in the pod are ready
```
Comment (yujuhong) [inline pkg/apis/core/types.go]: nit: end the comment with a period (`.`)

File: pkg/apis/core/types.go [hunk 7]
```diff
@@ -2409,6 +2411,12 @@ const (
 	TolerationOpEqual  TolerationOperator = "Equal"
 )
 
+// PodReadinessGate contains the reference to a pod condition
```
Comment (yujuhong) [inline pkg/apis/core/types.go]: Not sure why the `add ReadinessGates in pod spec` commit is in this PR...

**Review by yujuhong — COMMENTED** [review]
@freehan could you remove the readiness gate commits from this PR?

File: pkg/apis/core/types.go (hunk 7 above)
Comment (freehan) [reply pkg/apis/core/types.go]: Because it was not merged at the time...

File: staging/src/k8s.io/api/core/v1/types.go (hunk 2 above)
Comment (freehan) [reply staging/src/k8s.io/api/core/v1/types.go]: 27 is already taken

File: pkg/kubelet/kubelet_pods.go (hunk 3 above)
Comment (freehan) [reply pkg/kubelet/kubelet_pods.go]: sure

**Review by yujuhong — CHANGES_REQUESTED** [review]
Will review after the dependency is merged. Adding this to prevent the PR from merging.

**Comment by thockin** [comment]
Thanks @freehan and @yujuhong

**Comment by colemickens** [comment]
Hey folks, from the comments, it sounds like this is desired for 1.11, but it's not labeled correctly for that to happen, and there are additional labels needed if it doesn't merge today before the beginning of code freeze. Please ensure the labels for the milestone and urgency are set correctly.

**Comment by jberkus** [comment]
Going to bump to critical-urgent based on @freehan comment

/priority critical-urgent

**Review by yujuhong — APPROVED** [review]
/lgtm

**Comment by zparnold** [comment]
Hello there! @freehan I'm Zach Arnold working on Docs for the 1.11 release. This PR was identified as one needing some documentation in the https://github.com/kubernetes/website repo around your contributions (thanks by the way!) When you have some time, could you please modify/add/remove the relevant content that needs changing in our documentation repo? Thanks! Please let me or my colleague Misty know (@zparnold/@Misty on K8s Slack) if you need any assistance with the documentation.

**Comment by lucperkins** [comment]
@freehan Please note that I've created a PR for this in the website repo: https://github.com/kubernetes/website/pull/9197.

Please let me know if there'is any usage-related info that I should be aware of. Thus far I've added it to the list of possible Pod statuses but would love to add more than that.

**Comment by timchenxiaoyu** [comment]
we hava pod  container ready condition，i don't know， why we need  ContainersReady？

**Pull request merged**